File: src/lib/lang.ts
import { getLocale } from "./i18n";
import type { JobOptions, MediaInfo } from "./tauri";

export const COMMON_LANGUAGES: [string, string][] = [
  ["ger", "de"], ["eng", "en"], ["fre", "fr"], ["spa", "es"], ["ita", "it"], ["por", "pt"], ["dut", "nl"],
  ["pol", "pl"], ["rus", "ru"], ["ukr", "uk"], ["cze", "cs"], ["hun", "hu"], ["tur", "tr"], ["swe", "sv"],
  ["dan", "da"], ["nor", "no"], ["fin", "fi"], ["gre", "el"], ["jpn", "ja"], ["kor", "ko"], ["chi", "zh"],
  ["ara", "ar"], ["heb", "he"], ["hin", "hi"], ["tha", "th"], ["vie", "vi"],
];

const ALIASES: Record<string, string> = {
  deu: "ger", de: "ger", en: "eng", fra: "fre", fr: "fre", es: "spa", it: "ita", pt: "por", nld: "dut", nl: "dut",
  pl: "pol", ru: "rus", uk: "ukr", ces: "cze", cs: "cze", hu: "hun", tr: "tur", sv: "swe", da: "dan", no: "nor",
  nob: "nor", fi: "fin", ell: "gre", el: "gre", ja: "jpn", ko: "kor", zho: "chi", zh: "chi", ar: "ara", he: "heb",
  hi: "hin", th: "tha", vi: "vie",
};

export function normalizeLang(code: string): string {
  const c = code.trim().toLowerCase();
  return ALIASES[c] ?? c;
}

export function languageName(code: string | null | undefined): string {
  if (!code) return "";
  const canonical = normalizeLang(code);
  const two = COMMON_LANGUAGES.find(([c]) => c === canonical)?.[1] ?? (canonical.length === 2 ? canonical : null);
  if (!two) return code;
  try {
    return new Intl.DisplayNames([getLocale()], { type: "language" }).of(two) ?? code;
  } catch {
    return code;
  }
}

export interface SubSource {
  file: string | null;
  stream: number;
  codec: string;
  bitmap: boolean;
  language: string | null;
}

export function subtitleSources(media: MediaInfo, options: JobOptions): SubSource[] {
  const out: SubSource[] = media.subtitles.map((s, i) => ({ file: null, stream: i, codec: s.codec, bitmap: s.bitmap, language: s.language }));
  if (!options.skipExternalSubs) {
    for (const e of media.externalSubs ?? []) out.push({ file: e.path, stream: 0, codec: e.codec, bitmap: false, language: e.language });
  }
  for (const p of options.extraSubs ?? []) {
    if (out.some((s) => s.file === p)) continue;
    const parts = (p.split(/[\\/]/).pop() ?? "").split(".").slice(1, -1);
    const lang = parts.map(normalizeLang).find((x) => COMMON_LANGUAGES.some(([c]) => c === x)) ?? null;
    const ext = p.split(".").pop()?.toLowerCase();
    out.push({ file: p, stream: 0, codec: ext === "ass" || ext === "ssa" ? "ass" : ext === "vtt" ? "webvtt" : "subrip", bitmap: false, language: lang });
  }
  return out;
}

File: crates/zvc-core/src/lang.rs
const GROUPS: &[&[&str]] = &[
    &["ger", "deu", "de", "german", "deutsch"],
    &["eng", "en", "english"],
    &["fre", "fra", "fr", "french", "français", "francais"],
    &["spa", "es", "spanish", "español", "espanol"],
    &["ita", "it", "italian", "italiano"],
    &["por", "pt", "portuguese", "português", "portugues"],
    &["dut", "nld", "nl", "dutch", "nederlands"],
    &["pol", "pl", "polish", "polski"],
    &["rus", "ru", "russian"],
    &["ukr", "uk", "ukrainian"],
    &["cze", "ces", "cs", "czech"],
    &["slo", "slk", "sk", "slovak"],
    &["hun", "hu", "hungarian"],
    &["rum", "ron", "ro", "romanian"],
    &["gre", "ell", "el", "greek"],
    &["tur", "tr", "turkish"],
    &["swe", "sv", "swedish"],
    &["dan", "da", "danish"],
    &["nor", "nob", "nno", "no", "nb", "nn", "norwegian"],
    &["fin", "fi", "finnish"],
    &["ice", "isl", "is", "icelandic"],
    &["jpn", "ja", "japanese"],
    &["kor", "ko", "korean"],
    &["chi", "zho", "zh", "chinese"],
    &["ara", "ar", "arabic"],
    &["heb", "he", "iw", "hebrew"],
    &["hin", "hi", "hindi"],
    &["tha", "th", "thai"],
    &["vie", "vi", "vietnamese"],
    &["ind", "id", "indonesian"],
    &["may", "msa", "ms", "malay"],
    &["per", "fas", "fa", "persian"],
    &["hrv", "hr", "croatian"],
    &["srp", "sr", "serbian"],
    &["bul", "bg", "bulgarian"],
    &["slv", "sl", "slovenian"],
    &["est", "et", "estonian"],
    &["lav", "lv", "latvian"],
    &["lit", "lt", "lithuanian"],
    &["cat", "ca", "catalan"],
];

pub fn normalize(code: &str) -> String {
    let c = code.trim().to_lowercase();
    GROUPS
        .iter()
        .find(|g| g.contains(&c.as_str()))
        .map(|g| g[0].to_string())
        .unwrap_or(c)
}

pub fn matches(tag: Option<&str>, wanted: &str) -> bool {
    match tag {
        Some(t) if !t.trim().is_empty() && t != "und" => normalize(t) == normalize(wanted),
        _ => false,
    }
}

pub fn rank(tag: Option<&str>, wanted: &[String]) -> Option<usize> {
    wanted.iter().position(|w| matches(tag, w))
}

pub fn looks_like_code(s: &str) -> bool {
    let c = s.to_lowercase();
    (c.len() == 2 || c.len() == 3) && GROUPS.iter().any(|g| g.contains(&c.as_str()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn codes() {
        assert!(matches(Some("ger"), "de"));
        assert!(matches(Some("deu"), "German"));
        assert!(matches(Some("eng"), "en"));
        assert!(!matches(Some("eng"), "de"));
        assert!(!matches(Some("und"), "und"));
        assert!(!matches(None, "de"));
        assert_eq!(normalize("FRA"), "fre");
        assert_eq!(normalize("xx"), "xx");
        assert_eq!(rank(Some("eng"), &["de".into(), "en".into()]), Some(1));
        assert!(looks_like_code("de") && looks_like_code("eng") && !looks_like_code("forced") && !looks_like_code("x"));
    }
}
